runtime/src/accounts_cache.rs

```diff
@@ -0,0 +1,236 @@
+use dashmap::DashMap;
+use solana_sdk::{account::Account, clock::Slot, hash::Hash, pubkey::Pubkey};
+use std::{
+    collections::HashSet,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+};
+
+pub type SlotCache = Arc<SlotCacheInner>;
+
+#[derive(Default, Debug)]
+pub struct SlotCacheInner {
+    cache: DashMap<Pubkey, CachedAccount>,
+    same_account_writes: AtomicU64,
+    same_account_writes_size: AtomicU64,
+    unique_account_writes_size: AtomicU64,
+    is_frozen: AtomicBool,
+}
+
+impl SlotCacheInner {
+    pub fn report_slot_store_metrics(&self) {
+        datapoint_info!(
+            "slot_repeated_writes",
+            (
+                "same_account_writes",
+                self.same_account_writes.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "same_account_writes_size",
+                self.same_account_writes_size.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "unique_account_writes_size",
+                self.unique_account_writes_size.load(Ordering::Relaxed),
+                i64
+            )
+        );
+    }
+
+    pub fn insert(&self, pubkey: &Pubkey, account: Account, hash: Hash) {
+        if self.cache.contains_key(pubkey) {
+            self.same_account_writes.fetch_add(1, Ordering::Relaxed);
+            self.same_account_writes_size
+                .fetch_add(account.data.len() as u64, Ordering::Relaxed);
+        } else {
+            self.unique_account_writes_size
+                .fetch_add(account.data.len() as u64, Ordering::Relaxed);
+        }
+        self.cache.insert(*pubkey, CachedAccount { account, hash });
+    }
+
+    pub fn get_cloned(&self, pubkey: &Pubkey) -> Option<CachedAccount> {
+        self.cache
+            .get(pubkey)
+            // 1) Maybe can eventually use a Cow to avoid a clone on every read
+            // 2) Popping is only safe if its guaranteed only replay/banking threads
+            // are reading from the AccountsDb
+            .map(|account_ref| account_ref.value().clone())
+    }
+
+    pub fn mark_slot_frozen(&self) {
+        self.is_frozen.store(true, Ordering::SeqCst);
+    }
+
+    pub fn is_frozen(&self) -> bool {
+        self.is_frozen.load(Ordering::SeqCst)
```

Comment:
maybe, Ordering::Acquire is enough?

Reply:
In general I'm hesitant to use anything less than `SeqCst` for critical sections that are not well defined/fluid, because of maintainability. 

For instance, if we had another `atomic` that we expected to have a certain value by the time `is_frozen == true`, this can race because `StoreLoad` can be reordered in acquire-release semantics, see: https://stackoverflow.com/questions/41858540/whats-are-practical-example-where-acquire-release-memory-order-differs-from-seq for a classical example